src/libraries/System.Net.WebSockets/src/System/Net/WebSockets/ManagedWebSocket.KeepAlive.cs

```diff
@@ -0,0 +1,306 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Buffers;
+using System.Buffers.Binary;
+using System.Diagnostics;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.WebSockets
+{
+    internal sealed partial class ManagedWebSocket : WebSocket
+    {
+        private bool IsUnsolicitedPongKeepAlive => _keepAlivePingState is null;
+        private static bool IsValidSendState(WebSocketState state) => Array.IndexOf(s_validSendStates, state) != -1;
+        private static bool IsValidReceiveState(WebSocketState state) => Array.IndexOf(s_validReceiveStates, state) != -1;
+
+        private void HeartBeat()
+        {
+            if (IsUnsolicitedPongKeepAlive)
+            {
+                UnsolicitedPongHeartBeat();
+            }
+            else
+            {
+                KeepAlivePingHeartBeat();
+            }
+        }
+
+        private void UnsolicitedPongHeartBeat()
+        {
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this);
+
+            Observe(
+                TrySendKeepAliveFrameAsync(MessageOpcode.Pong));
+        }
+
+        private ValueTask TrySendKeepAliveFrameAsync(MessageOpcode opcode, ReadOnlyMemory<byte>? payload = null)
+        {
+            Debug.Assert(opcode is MessageOpcode.Pong || !IsUnsolicitedPongKeepAlive && opcode is MessageOpcode.Ping);
+
+            if (!IsValidSendState(_state))
+            {
+                if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this, $"Cannot send keep-alive frame in {nameof(_state)}={_state}");
+
+                // we can't send any frames, but no need to throw as we are not observing errors anyway
+                return ValueTask.CompletedTask;
+            }
+
+            payload ??= ReadOnlyMemory<byte>.Empty;
+
+            return SendFrameAsync(opcode, endOfMessage: true, disableCompression: true, payload.Value, CancellationToken.None);
+        }
+
+        private void KeepAlivePingHeartBeat()
+        {
+            Debug.Assert(_keepAlivePingState != null);
+
+            if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this);
+
+            bool shouldSendPing = false;
+            long pingPayload = -1;
+
+            try
+            {
+                lock (StateUpdateLock)
+                {
+                    if (_keepAlivePingState.Exception is not null)
+                    {
+                        if (NetEventSource.Log.IsEnabled()) NetEventSource.Trace(this, $"KeepAlive already faulted, skipping... (exception: {_keepAlivePingState.Exception.Message})");
+                        return;
+                    }
+
+                    long now = Environment.TickCount64;
```

Comment:
This looks unused

Reply:
Presumably the two uses of `Environment.TickCount64` below should be replaced by `now`?